File: 1130.java
class Solution {
    public int mctFromLeafValues(int[] arr) {
        Stack<Integer> stack = new Stack<>();
        stack.push(Integer.MAX_VALUE);
        int ans = 0;
        for (int i = 0; i < arr.length; i++){
            while (stack.peek() < arr[i])
                ans += stack.pop()*Math.min(arr[i],stack.peek());
            stack.push(arr[i]);
        }

        while(stack.size() > 2)
            ans += stack.pop()*stack.peek();
        
        return ans;
    }
}

File: 1130.cpp
class Solution {
public:
    int mctFromLeafValues(vector<int>& arr) {
        int res = 0;
        stack<int> s;
        s.push(INT_MAX);
        int n = arr.size();
        for (auto& a : arr) {
            while (s.top() <= a) {
                int temp = s.top();
                s.pop();
                res += temp * min(s.top(), a);
            }
            s.push(a);
        }
        while (s.size() > 2) {
            int temp = s.top();
            s.pop();
            res += temp * s.top();
        }
        return res;
    }
};